### PR TITLE
Expose JSX compilation to renderers

### PR DESCRIPTION
**Pull request by natemoo-re**
> This is slated for `astro@0.18.0`!

## Terminology

- "JSX" is used here to refer to both `.jsx` and `.tsx` files. They are conceptually interchangeable.

## Motivation

As it turns out, JSX is really complicated! Most JSX approaches (rightly) assume that your entire project is full of components that should be compiled the same way. Astro is in a unique position by promising that users can **mix components** from different frameworks together—what is simple for React + Vue + Svelte becomes very complicated for JSX-based frameworks like React + Preact + Solid. 

> The only reason Astro currently allows React + Preact is because Snowpack smartly detects `import { h } from 'preact'`, defaulting to React otherwise. We can't scale this approach.

At the same time, JSX is a common syntax without universal agreement on how compilation should work. The most common approach is to use `jsxFactory` (`React.createElement` or `h`) and `jsxFragment` (`React.Fragment` or `Fragment`), and [this is all esbuild supports](https://github.com/evanw/esbuild/issues/334). Frameworks are moving away from this approach and [tools are moving to support these more complex JSX configurations](https://github.com/alloc/vite-react-jsx).

React recently introduced a [new JSX Transform](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html). Vue takes [yet another approach](https://github.com/vuejs/jsx-next). Solid [compiles JSX](https://github.com/solidjs/solid/tree/main/packages/babel-preset-solid) without any VDOM implementation (similar to Svelte) and expects different output in SSR mode than on the client.

## Goals

- [x] Find a way for renderers to take granularly take responsibility for a JSX file, to enable per-file JSX compilation.
- [x] Support non-standard JSX compilation that does not use a simple global pragma.
- [x] Support JSX compilation that should behave differently for SSR versus client
- [x] **Stretch** Users expect to _not_ `import { jsxPragma } from 'lib'`. Can we enable this for convenience?
- [ ] **Stretch** While we're taking responsibility for compiling JSX, can we automatically enable tools like [`react-refresh`](https://www.npmjs.com/package/react-refresh) and [`prefresh`](https://github.com/preactjs/prefresh)?

## Solution

At a high-level, this PR does a few things:
- **Introduces a Snowpack Plugin to `load` `.jsx` and `.tsx` files.**
- **Allows renderers to manage JSX compilation on a per-file basis.** `es-module-lexer` scans the loaded file for imports and matches that to the renderer. Files with no imports may use a `/** @jsxImportSource my-library */` comment.
- **Allows renderers to compile JSX files.** Renderers can control how a JSX file is transformed and are passed the `load` context (`isSSR`, `isDev`, `isHmrEnabled`, etc). Babel is the de facto solution for this—every framework has a Babel plugin. Therefore, renderers may configure which Babel plugins should be used to process the JSX file.

### `@babel/core` is slow. We should use `esbuild`!

`esbuild` is awesome, but it only supports `jsxFactory` and `jsxFragment`. Vite enables automatic imports with the work-around [`jsxInject`](https://vitejs.dev/guide/features.html#jsx) option, which Snowpack introduced as well. This doesn't solve the core problem here, which is that **not every JSX file uses jsxFactory**!

Every framework supports Babel. In practice, every JSX file in Snowpack is likely being run through Babel already! [`@snowpack/plugin-react-refresh`](https://github.com/snowpackjs/snowpack/tree/main/plugins/plugin-react-refresh) and [`@prefresh/snowpack`](https://github.com/preactjs/prefresh/blob/main/packages/snowpack/src/index.js#L85-L98) both use it for `transform`. Any users of other frameworks are likely using a custom Babel config and relying on [`@snowpack/plugin-babel`](https://github.com/snowpackjs/snowpack/tree/main/plugins/plugin-babel).

This PR also attempts to mitigate Babel's slowness as much as possible. `esbuild` pre-compiles the file, transforming most non-standard syntax and removing TypeScript—the only non-standard syntax preserved is JSX. So Babel's only responsibility here is to compile JSX to normal JS, which should be relatively fast. 

There is precedent for this in the linked Snowpack plugins above, and those haven't been a particular performance concern. Babel _can_ be fast if you keep it focused.

### Multi-framework support is an edge case. Multi-JSX-framework support is and edge case within that edge case.

Multi-framework is probably a 20% use case feature of Astro, but it is a very important conceptual one. This PR makes every attempt to skip work when only a single renderer is enabled, which we can guess is about 80% of the time. This PR's more complex scanning behavior will kick-in only if multiple JSX-supporting renderers are enabled.

At the same time, this PR actually makes life _easier_ for users of a single framework! They won't need to configure any JSX compilation using `snowpack.config.mjs` _or_ manually import their JSX factory function. Again, Preact is special-cased in Snowpack so we're not seeing that pain for users of smaller frameworks. The JSX factory import has tripped up quite a few users coming from Next.js (Jason, Cassidy, etc).

Finally, mixing multiple JSX frameworks shouldn't be any more difficult than mixing React + Vue + Svelte! It's the same feature. Astro should own that complexity, not the user.

> Imagine how great it would be for React users to test out Solid on a page or two without any extra hassle. Just add `@astrojs/renderer-solid` and you're good to go.

### We shouldn't deal with Fast Refresh here!

Maybe, but shouldn't we if it's a small lift? It's clear that HMR is table stakes for dev tooling, but Fast Refresh (aka stateful HMR) is increasingly moving that way, too. The "fast refresh" plugins for Snowpack are just running a file through Babel anyway, so this would remove the need for those _and_ make users super happy.

## Testing

- [ ] TODO: add tests

## Docs

- [x] TODO: add docs

### CODE REVIEW

**Comment by matthewp** [comment]
Before you go too far with this, this seems like a heavy solution. I don't think we want to add babel processing to every `.jsx` file if we can avoid it. This also runs esbuild multiple times, I'm sure there's a reason but that is going to be a performance issue.

Can we not just prepend the appropriate import statement to the file if its missing?

**Comment by natemoo-re** [comment]
@matthewp You're right that this is a pretty heavy solution, but there are very good reasons for handling it this way! I'll take some time to write up a good outline before I write any more code. It should make more sense in context.

That being said, the double `esbuild` transform is a hack and I'm planning to remove it soon.

**Comment by natemoo-re** [comment]
cc @eyelidlessness, I know this was one blocker for the Solid renderer.

**Comment by jasikpark** [comment]
Would SWC be a viable way to pull in babel transforms? https://swc.rs/docs/comparison-babel/ -- it's written in Rust, so it should be fast similar to esbuild...

**Comment by natemoo-re** [comment]
Was able to build 98% of a Solid renderer based on this branch, which was the main motivation here. [Check out the diff](https://github.com/snowpackjs/astro/compare/feat/improve-jsx...feat/renderer-solid).

**Comment by natemoo-re** [comment]
> Would SWC be a viable way to pull in babel transforms? https://swc.rs/docs/comparison-babel/ -- it's written in Rust, so it should be fast similar to esbuild...

Ahh. Correct me if I'm wrong, but I think that link just compares SWC's featureset to the most common babel plugins. It doesn't look like SWC can actually run babel plugins.

**Comment by jasikpark** [comment]
I suppose you're right, since otherwise that would probably be listed in https://swc.rs/docs/usage-plugin rather than describing only how to make `swc` specific plugins.

**Comment by jasikpark** [comment]
https://github.com/swc-project/swc/pull/1465 It does seem like they are working on babel plugin support though 🤔 I'll look into it more.

File: packages/astro/snowpack-plugin-jsx.cjs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  const {
+    configManager
+  } = options;
+
+  let didInit = false;
+  return {
+    name: '@astrojs/snowpack-plugin-jsx',
+    resolve: {
+      input: ['.jsx', '.tsx'],
+      output: ['.js'],
```
Comment (natemoo-re) [inline packages/astro/snowpack-plugin-jsx.cjs:39]: We're taking over resolution of `.jsx` and `.tsx` files from Snowpack.

File: packages/astro/snowpack-plugin-jsx.cjs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        await eslexer.init;
+        didInit = true;
+      }
+
+      const contents = await fs.readFile(filePath, 'utf8');
+      const loader = getLoader(fileExt);
+
+      const { code, warnings } = await esbuild.transform(contents, {
+        loader,
+        jsx: 'preserve',
```
Comment (natemoo-re) [inline packages/astro/snowpack-plugin-jsx.cjs:52]: We transform with `esbuild` BUT preserve the JSX. This brings us from `TSX` (or non-standard JS) => standard `JSX`.

File: packages/astro/snowpack-plugin-jsx.cjs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+      const importSources = new Set(renderers.map(({ jsxImportSource }) => jsxImportSource).filter(i => i));
+      const getRenderer = (importSource) => renderers.find(({ jsxImportSource }) => jsxImportSource === importSource);
+      const getTransformOptions = async (importSource) => {
+        const { name } = getRenderer(importSource);
+        const { default: renderer } = await import(name);
+        return renderer.jsxTransformOptions(transformContext);
+      }
+
+      // If we only have a single renderer, we can skip a bunch of work!
+      if (importSources.size === 1) {
```
Comment (natemoo-re) [inline packages/astro/snowpack-plugin-jsx.cjs:73]: If only one of our renderers cares about JSX files, we can just use that one!

File: packages/astro/snowpack-plugin-jsx.cjs [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        const result = transform(code, filePath, await getTransformOptions(Array.from(importSources)[0]))
```
Comment (natemoo-re) [inline packages/astro/snowpack-plugin-jsx.cjs:74]: Transform the (now plain) JSX file with the renderer's Babel options. This allows renderers to conditionally modify Babel config based on SSR/HMR/etc.

File: packages/astro/snowpack-plugin-jsx.cjs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        return {
+          '.js': {
+            code: result.code || ''
+          },
+        };
+      }
+
+      // we need valid JS to scan for imports
+      // so let's just use `h` and `Fragment` as placeholders
+      const { code: codeToScan } = await esbuild.transform(code, {
```
Comment (natemoo-re) [inline packages/astro/snowpack-plugin-jsx.cjs:85]: Unfortunately, if we have multiple renderers, we need to scan their imports. This means one more pass through `esbuild` to get valid JS (no JSX). The result doesn't need to run, it will be discarded.

File: packages/astro/snowpack-plugin-jsx.cjs [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        jsxFactory: 'h',
+        jsxFragment: 'Fragment',
+      });
+
+      const [imports] = eslexer.parse(codeToScan);
+      let importSource;
+
+      if (imports) {
+        for (let { n: name } of imports) {
+          if (name.indexOf('/') > -1) name = name.split('/')[0];
```
Comment (natemoo-re) [inline packages/astro/snowpack-plugin-jsx.cjs:97]: Scan for package specifiers that match the renderer's `jsxImportSource`. These can be identical OR sub-packages like `preact/hooks`.

File: packages/astro/snowpack-plugin-jsx.cjs [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+            break;
+          }
+        }
+      }
+
+      if (!importSource) {
+        let match;
+
+        while ((match = /\/\*\*(?:[^*][^/]|\s)*@jsxImportSource\s+(.+)\s*\*\//gm.exec(contents)) !== null) {
+          importSource = match[1].trim();
```
Comment (natemoo-re) [inline packages/astro/snowpack-plugin-jsx.cjs:109]: If we didn't find a matching import for any of our renderers, maybe this file doesn't have any imports! Users can add a 
```
/** @jsxImportSource preact */
```
comment to ensure a file gets run through the correct renderer.

File: packages/astro/snowpack-plugin-jsx.cjs [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+          break;
+        }
+      }
+
+      if (!importSource) {
+        console.log(`${filePath}
+Unable to resolve JSX transformer! If you have more than one renderer enabled, you should use a pragma comment.
+/* jsxImportSource: preact */
+`);
```
Comment (natemoo-re) [inline packages/astro/snowpack-plugin-jsx.cjs:118]: We still haven't matched a renderer, so let's give users a hint that they should use a pragma comment.

File: packages/astro/snowpack-plugin-jsx.cjs [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        return {
+          '.js': {
+            code: ''
+          },
+        }
+      }
+
+      const result = transform(code, filePath, await getTransformOptions(importSource));
```
Comment (natemoo-re) [inline packages/astro/snowpack-plugin-jsx.cjs:126]: Finally! We matched `jsxImportSource` and can run this file through Babel.

File: packages/astro/snowpack-plugin-jsx.cjs [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    presets,
+    plugins: [...plugins, alias ? ['babel-plugin-module-resolver', { root: process.cwd(), alias }] : undefined].filter(v => v),
+    cwd: process.cwd(),
+    filename: id,
+    ast: false,
+    compact: false,
+    sourceMaps: false,
+    configFile: false,
+    babelrc: false,
+  });
```
Comment (natemoo-re) [inline packages/astro/snowpack-plugin-jsx.cjs:155]: This opts out of all Babel's weird auto-checking.

File: packages/create-astro/src/frameworks.ts [hunk 11]
```diff
@@ -36,6 +35,25 @@ export default function ReactCounter({ children }) {
...
+  const subtract = () => setCount(count() - 1);
+
+  return (
+    <div id="solid" class="counter">
+      <button onClick={subtract}>-</button>
+      <pre>{count()}</pre>
+      <button onClick={add}>+</button>
+    </div>
+  );
+}
```
Comment (natemoo-re) [inline packages/create-astro/src/frameworks.ts:56]: Added Solid to `create-astro`!

File: packages/renderers/renderer-preact/index.js [hunk 12]
```diff
@@ -2,5 +2,14 @@ export default {
...
   client: './client',
   server: './server',
-  knownEntrypoints: ['preact', 'preact-render-to-string'],
+  knownEntrypoints: ['preact', 'preact/jsx-runtime', 'preact-render-to-string'],
+  jsxImportSource: 'preact',
+  jsxTransformOptions: async () => {
+    const { default: { default: jsx }} = await import('@babel/plugin-transform-react-jsx');
+    return {
+      plugins: [
+        jsx({}, { runtime: 'automatic', importSource: 'preact' })
```
Comment (natemoo-re) [inline packages/renderers/renderer-preact/index.js:11]: This is the new JSX transform that uses `preact/jsx-runtime`!

File: packages/renderers/renderer-preact/server.js [hunk 13]
```diff
@@ -10,7 +10,15 @@ function check(Component, props, children) {
...
-  return typeof html === 'string';
+
+  if (typeof html !== 'string') {
+    return false;
+  }
+
+  // There are edge cases (SolidJS) where Preact *might* render a string,
+  // but components would be <undefined></undefined>
+
+  return !/\<undefined\>/.test(html);
```
Comment (natemoo-re) [inline packages/renderers/renderer-preact/server.js:21]: Solid components can run through Preact's `renderToString` fine! But they return `<undefined>`... So that's a weird case to check for here.

File: packages/renderers/renderer-react/index.js [hunk 14]
```diff
@@ -2,5 +2,14 @@ export default {
...
   client: './client',
   server: './server',
-  knownEntrypoints: ['react', 'react-dom', 'react-dom/server'],
+  knownEntrypoints: ['react', 'react/jsx-runtime', 'react-dom', 'react-dom/server.js'],
+  jsxImportSource: 'react',
+  jsxTransformOptions: async () => {
+    const { default: { default: jsx }} = await import('@babel/plugin-transform-react-jsx');
+    return {
+      plugins: [
+        jsx({}, { runtime: 'automatic', importSource: 'react' })
```
Comment (natemoo-re) [inline packages/renderers/renderer-react/index.js:11]: This is the new JSX transform that uses `react/jsx-runtime`!

File: packages/renderers/renderer-solid/client.js [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+    ...props,
+    children: element.querySelector('astro-fragment'),
+  });
+
+  const children = Array.isArray(component)
+    ? component
+    : [ component ];
+
+  element.replaceChildren(...children);
+}
```
Comment (natemoo-re) [inline packages/renderers/renderer-solid/client.js:17]: Huge thanks to @eyelidlessness for the help here!

File: packages/renderers/renderer-solid/index.js [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+        solid({}, { generate: isSSR ? 'ssr' : 'dom' }),
+      ]
+    }
+
+    if (isSSR) {
+      options.alias = {
+        'solid-js/web': 'solid-js/web/dist/server.js',
+        'solid-js': 'solid-js/dist/server.js',
+      };
+    }
```
Comment (natemoo-re) [inline packages/renderers/renderer-solid/index.js:21]: If we're in SSR mode, we need to specify how these should be resolved so that Snowpack doesn't grab the wrong packages for us.

File: packages/renderers/renderer-solid/index.js [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+  client: './client',
+  server: './server',
+  knownEntrypoints: ['solid-js', 'solid-js/web'],
+  external: ['solid-js/web/dist/server.js', 'solid-js/dist/server.js', 'babel-plugin-module-resolver', 'babel-preset-solid'],
+  jsxImportSource: 'solid-js',
+  jsxTransformOptions: async ({ isSSR }) => {
+    const [{ default: solid }] = await Promise.all([import('babel-preset-solid')]);
+    const options = {
+      presets: [
+        solid({}, { generate: isSSR ? 'ssr' : 'dom' }),
```
Comment (natemoo-re) [inline packages/renderers/renderer-solid/index.js:12]: `generate` is different based on whether it's SSR or not!

**Review by natemoo-re — COMMENTED** [review]
Leaving some notes! Largely documentation changes but I called out the important stuff (at least from my perspective).

File: packages/astro/snowpack-plugin-jsx.cjs (hunk 5 above)
Comment (FredKSchott) [reply packages/astro/snowpack-plugin-jsx.cjs:96]: I'm really -1 on a build failing when imports change, basically the user story of:

1. I refactor a file of React components to remove the need for `useState`
2. I clean up my imports, removing `import {useState} from 'react';`
3. my build fails because the imports changed.

Can we just be more explicit, and always use the first matched renderer in your config file _unless_ the `jsxImportSource` pragma comment is used?

I know we _can_ be smart than this with the imports, but I think it leads to more surprises (the bad kind) in practice.

File: packages/renderers/renderer-preact/index.js (hunk 12 above)
Comment (FredKSchott) [reply packages/renderers/renderer-preact/index.js:11]: nice!

File: packages/renderers/renderer-preact/server.js (hunk 13 above)
Comment (FredKSchott) [reply packages/renderers/renderer-preact/server.js:21]: very strange.

Can this idea of "check()" now go away? It sounds like we have added the ability to know which framework a file was transformed for. If we could surface that info somehow, then we wouldn't need this runtime check...

I'm okay if this is a follow up PR, but maybe its actually really easy?

**Review by FredKSchott — COMMENTED** [review]
Looking really good! Still needs tests :)

File: packages/astro/snowpack-plugin-jsx.cjs (hunk 5 above)
Comment (matthewp) [reply packages/astro/snowpack-plugin-jsx.cjs:96]: @FredKSchott that refactor hazard has existed as long as JSX has though. I think requiring people add a non-standard comment to every file is asking quite a lot. The mental model of a JSX expression being syntactic sugar for a function call makes sense. Now making it an external config value feels like a regression to me.

Comment (natemoo-re) [reply packages/astro/snowpack-plugin-jsx.cjs:96]: @matthewp Agreed, but I would point out that the comment is borrowed from [`@babel/plugin-transform-react-jsx`](https://babeljs.io/docs/en/babel-plugin-transform-react-jsx#customizing-the-automatic-runtime-import).

@FredKSchott I'm wary of having the order of the renderers in the config array impact the behavior in an unintuitive way. I've noticed that trip a lot of people up with other tools. I'm also wary of adding another property like `fallbackRenderer` that only applies for JSX. Can we solve this with a good error message?


```
Unable to resolve JSX transformer! With more than one renderer enabled, you should use a pragma comment or include an import.
Add `/* jsxImportSource: preact */` or `import { h } from "preact"`
```

Comment (FredKSchott) [reply packages/astro/snowpack-plugin-jsx.cjs:96]: This just feels a lot like we're optimizing for the 1% use case at the expense of the 99% use case. If most users are only ever using one framework, they're going to see this warning even though they only have one JSX renderer connected.

How about:
- If the user only has 1 JSX renderer defined in config, default to that
- If the user has 2+ JSX renderers defined in config, show the warning from the comment above

Comment (FredKSchott) [reply packages/astro/snowpack-plugin-jsx.cjs:96]: Also if we're word-smithing the warning, I'd say:

```
Unable to detect JSX framework! Add `/* jsxImportSource: preact */` to the file to fix.
```

Comment (matthewp) [reply packages/astro/snowpack-plugin-jsx.cjs:96]: A user that only has one framework won't run into this issue, the code always uses the first import source, unless I'm misreading this: https://github.com/snowpackjs/astro/pull/588/files#diff-4326d59c6d077cb5cc64b38132f64e52a47473932d71225703a0fae2fe6b59adR84

I also disagree with the notion that framework swapping is going to be rare. I think that's y

Comment (matthewp) [reply packages/astro/snowpack-plugin-jsx.cjs:96]: A user with one framework won't run into the import scanning because it's already using the first importSource in that case, unless I'm misunderstanding this code: https://github.com/snowpackjs/astro/pull/588/files#diff-4326d59c6d077cb5cc64b38132f64e52a47473932d71225703a0fae2fe6b59adR84

Comment (FredKSchott) [reply packages/astro/snowpack-plugin-jsx.cjs:96]: doh, I missed that and I even commented on that line! If that's already what you're proposing @natemoo-re then I'm on board 👍

Comment (natemoo-re) [reply packages/astro/snowpack-plugin-jsx.cjs:101]: That's correct! This only ever warns if a user is using multiple JSX-aware renderers. I also improved the warning to suggest the user's first configured renderer automatically.

File: packages/renderers/renderer-preact/server.js (hunk 13 above)
Comment (natemoo-re) [reply packages/renderers/renderer-preact/server.js:21]: I'll have to think about this! The way it's structured right now, `check` is a runtime concern, whereas this compilation is a build-time concern. 

There might be a clean way to refactor this, but I think that's out of scope for this PR.

**Comment by natemoo-re** [comment]
Tests should pass once https://github.com/snowpackjs/snowpack/pull/3602 lands and gets merged back in

**Review by FredKSchott — APPROVED** [review]
LGTM once the tests pass!
